src/lib/zkapp.ts

```diff
@@ -93,84 +110,277 @@ function method<T extends SmartContract>(
+            let blindingValue = getBlindingValue();
+            let [{ memoized }, result] = memoizationContext.runWith(
+              {
+                memoized: [],
+                currentIndex: 0,
+                blindingValue,
+              },
+              () => {
+                method.apply(this, actualArgs);
+              }
+            );
+            assertStatePrecondition(this);
+
+            // connect our input + result with callData, so this method can be called
+            let callDataFields = methodArgumentsToFields(
+              methodIntf,
+              actualArgs
+            );
+            if (methodIntf.returnType) {
+              callDataFields.push(...methodIntf.returnType.toFields(result));
+            }
+            callDataFields.push(getMethodId(methodIntf, methodIndex));
+            callDataFields.push(blindingValue);
+            party.body.callData = Poseidon.hash(callDataFields);
+
+            if (!Authorization.hasAny(party)) {
+              Authorization.setLazyProof(party, {
+                methodName: methodIntf.methodName,
+                args: clonedArgs,
+                // proofs actually don't have to be cloned
+                previousProofs: getPreviousProofsForProver(
+                  actualArgs,
+                  methodIntf
+                ),
+                ZkappClass,
+                memoized,
+                blindingValue,
+              });
+            }
+            return result;
+          }
         }
-      );
-      return result;
-    } else if (!Mina.currentTransaction.has()) {
-      // outside a transaction, just call the method, but check precondition invariants
-      let result = method.apply(this, actualArgs);
-      // check the self party right after calling the method
-      // TODO: this needs to be done in a unified way for all parties that are created
-      assertPreconditionInvariants(this.self);
-      cleanPreconditionsCache(this.self);
-      assertStatePrecondition(this);
-      return result;
-    } else {
-      // in a transaction, also add a lazy proof to the self party
-      // (if there's no other authorization set)
-
-      // first, clone to protect against the method modifying arguments!
-      // TODO: double-check that this works on all possible inputs, e.g. CircuitValue, snarkyjs primitives
-      let clonedArgs = cloneCircuitValue(actualArgs);
-      let result = method.apply(this, actualArgs);
-      assertStatePrecondition(this);
-      let party = this.self;
-      if (!Authorization.hasAny(party)) {
-        Authorization.setLazyProof(party, {
-          method,
-          args: clonedArgs,
-          // proofs actually don't have to be cloned
-          previousProofs: getPreviousProofsForProver(actualArgs, methodIntf),
-          ZkappClass,
-        });
-      }
-      return result;
+      )[1];
     }
+    // if we're here, this method was called inside _another_ smart contract method
```

Comment:
this is what happens inside the caller, _inside_ calling another method. the code below has some comments to explain what's happening